flang/lib/Lower/OpenMP.cpp

```diff
@@ -178,9 +179,19 @@ genOMP(Fortran::lower::AbstractConverter &converter,
     // Ensure the block is well-formed.
     firOpBuilder.create<mlir::omp::TerminatorOp>(currentLocation);
 
-    // Place the insertion point to the start of the first block.
-    firOpBuilder.setInsertionPointToStart(&block);
-  }
+      // Place the insertion point to the start of the first block.
+      firOpBuilder.setInsertionPointToStart(&block);
+    } else if (parallelDirective.v == llvm::omp::OMPD_master) {
+      auto masterOp =
```

Comment:
I guess it would make sense to have a function that create this op with block and terminator. Something like: https://github.com/flang-compiler/f18-llvm-project/blob/870827f65222ebb1a9701d4d0e916f315dc692a8/flang/lib/Lower/OpenACC.cpp#L70

Reply:
Same approach can't be used in here, since we(OpenMP) uses different builders. However thanks for the suggestion I've tried to factor out the common code out of these `Op` creation and put in a helper(`finalizeRegionOp`). Fits well here!

Reply:
Right. Looks fine like that.

Reply:
Nit: The body is also created in finalizeRegionOp? If so would createBody be a better name?